### ISSUE TITLE
Stop publishing images for jupyterhub

### ISSUE THREAD

**Issue by minrk** [jupyterhub/jupyterhub#4954]
### Proposed change

Related to https://github.com/jupyterhub/team-compass/issues/474 I think maybe we should stop publishing jupyterhub images altogether.

We currently publish these images from this repo:

- jupyterhub/demo - a toy complete environment for quickstart/testing purposes (not useful for any deployment) where users are run in the container
- jupyterhub/jupyterhub - includes hub and proxy, no Spawner cannot run users  - not useful for production
- jupyterhub/jupyterhub-onbuild (adds a single `ONBUILD ADD` to jupyterhub/jupyterhub) for a minor convenience as a base image
- jupyterhub/singleuser - shouldn't be used, offers nothing over docker-stacks

Production deployments do not generally use these images because:

- they do not include a Spawner, so can't be used on their own, they can only be a _base_ image for other custom images
- the base image part can be fully replaced with a single `pip install` or `conda install` in a more nontrivial base image
- jupyterhub/jupyterhub includes CHP, which should not generally be run in the same image, and we shouldn't be implicitly encouraging it

Maintaining these images on this repo causes maintenance problems like #4953  and the various motivations for https://github.com/jupyterhub/team-compass/issues/474

I propose that we stop publishing these images altogether until/unless we get image-building going from its own repo. I wish I had proposed this before 5.0, since a major version is a nice cutoff for a change in practice, but the second best time is now.

The only image I think we _might_ keep is `demo`, since it serves a unique purpose and is clearly not for production use. The others may give the _impression_ of being appropriate for something they shouldn't really be used for.

Stopping publishing from here does not _preclude_ us from publishing images (even these same images) from another repo. But it does mean that we need to set up infrastructure _dedicated to building and maintaining images_ before we start building these images again, which I think will help us be thoughtful about it.

### Alternative options

- Keep publishing  from this repo (I don't think we should do this, but it will keep happening if we don't change something)
- Move image-building to its own repo in a more thoughtful way _before_ we stop building here, focused on building images that address real use cases

### Who would use this feature?

The real question is: is anyone really using `jupyterhub/jupyterhub` who would be inconvenienced by this? `jupyterhub/singleuser` is an empty layer on top of `base-notebook`. `quay.io/jupyterhub/jupyterhub` is pulled over 1000 times every day, but looking at the logs quay.io gives us, about 80% of all worldwide pulls of jupyterhub/jupyterhub come from a single instance of [watchtower](https://containrrr.dev/watchtower/) in the Netherlands, which has pulled it over 10,000 times in the last week.

**Comment by manics** [comment]
I think one of the reasons https://github.com/jupyterhub/team-compass/issues/474 stalled was it tries to do everything. If we scoped this to just creating a single repo with GitHub workflows whose only purpose is to publish the existing JupyterHub images that should be doable, and means we won't break anyone relying on these images.

**Comment by manics** [comment]
Here's what I'm thinking for the new container repo:

- All future JupyterHub image tags must be of the form `quay.io/jupyterhub/image:MAJOR.MINOR.PATCH-BUILDNUMBER`
- We use `requirements.txt` and `requirements.in` with dependabot to manage the JupyterHub version used when building the images
- We _don't_ use GitHub tags, instead every PR leads to a tagged image. This is automated by:
  - Extract the JupyterHub `{version}` from requirements.txt
  - Lookup all quay.io tags of the form `quay.io/jupyterhub/image:{version}-BUILDNUMBER` using the quay.io API, autoincrement `BUILDNUMBER`
    - I've started work on a GitHub action to do this: https://github.com/manics/action-get-quayio-tags/blob/main/src/buildNumbers.js
  - Automatically tag the new image as `quay.io/jupyterhub/image:MAJOR.MINOR.PATCH-BUILDNUMBER+1`
  - Rely on https://github.com/jupyterhub/action-major-minor-tag-calculator/ to automatically create the additional tag aliases `MAJOR.MINOR.PATCH`, MAJOR.MINOR`, `MAJOR` as at present.

**Comment by manics** [comment]
I think https://github.com/manics/jupyterhub-container-images is almost ready

https://github.com/manics/jupyterhub-container-images/tree/3b1f9ad597b8427c6a3f6b25c884d932731d7ff0 is the result of `git-filter-repo --path Dockerfile --path singleuser/ --path demo-image/ --path onbuild/ --path .github/workflows/release.yml` on https://github.com/jupyterhub/jupyterhub/tree/5.2.1

These are the changes I've made on top of that https://github.com/manics/jupyterhub-container-images/compare/3b1f9ad597b8427c6a3f6b25c884d932731d7ff0...main

It uses
- https://github.com/manics/action-get-quayio-tags
- https://github.com/jupyterhub/action-major-minor-tag-calculator/pull/354

It's currently pushing to
- https://quay.io/repository/manics/jupyterhub-image-test?tab=tags
- https://quay.io/repository/manics/jupyterhub-image-test-onbuild?tab=tags
- https://quay.io/repository/manics/jupyterhub-image-test-demo?tab=tags
- https://quay.io/repository/manics/jupyterhub-singleuser-image-test?tab=tags

**Comment by manics** [comment]
Our [docker build CI workflow](https://github.com/jupyterhub/jupyterhub/actions/workflows/release.yml) has broken. I haven't looked closely at the current problem, but it's probably time to focus on getting the build decoupled! First step is updating (or creating a new action/library) to allow quay.io tags to feed in to our image tag calculator.
https://github.com/jupyterhub/action-major-minor-tag-calculator/pull/354

**Comment by minrk** [comment]
@manics do you think you'll have time to set up the new repo before we're ready for a release (at least a month)? If so, I'll remove the docker-publish workflow now so we don't have to worry about fixing it. I'm happy to review the new setup, if you point me to where/when that would be useful. Worst case, we can revert the removal and fix it then.

**Comment by manics** [comment]
Yes, I should be able to.

**Comment by manics** [comment]
It's been a month and I haven't made any significant changes.... so shall we go for it? Steps to getting this into production (in no particular order, and we could push some of these to later):

- [ ] Review https://github.com/manics/jupyterhub-container-images, in particular the GitHub workflow https://github.com/manics/jupyterhub-container-images/blob/main/.github/workflows/build.yml
- [ ] Transfer https://github.com/manics/jupyterhub-container-images to the JupyterHub GitHub org
- [ ] Transfer https://github.com/manics/action-get-quayio-tags to the JupyterHub GitHub org
- [ ] Change `env.IMAGE` and `env.SINGLEUSER` image names to `jupyterhub./` in https://github.com/manics/jupyterhub-container-images/blob/main/.github/workflows/build.yml
- [ ] Change `env.PUBLISH_DOCKERIO` to `true` in https://github.com/manics/jupyterhub-container-images/blob/main/.github/workflows/build.yml to publish to docker hub (in addition to quay.io)
- [ ] Create repository secrets `secrets.QUAY_USERNAME` `secrets.QUAY_PASSWORD` `secrets.DOCKERHUB_USERNAME` `secrets.DOCKERHUB_TOKEN`
- [ ] Review which images we support (e.g. I think there's a suggestion somewhere that we remove the `onbuild` image?)
- [ ] Test the built images, we're managing a requirements.txt file https://github.com/manics/jupyterhub-container-images/blob/main/base/requirements.txt which means we could bump a dependency to a new version which breaks JupyterHub

**Comment by minrk** [comment]
Nice! What would be the process for doing e.g. a security patch for the 5.x image once 6.x is being published? Just merge into a PR to a 5.x branch?


> - Transfer https://github.com/manics/jupyterhub-container-images to the JupyterHub GitHub org
> - Transfer https://github.com/manics/action-get-quayio-tags to the JupyterHub GitHub org

Having given those a quick look, I think you can go ahead and move those now.


> Review which images we support (e.g. I think there's a suggestion somewhere that we remove the onbuild image?)

I think we should drop onbuild. It's just not worth maintaining for the savings of a single implicit `COPY jupyterhub_config.py /srv/jupyterhub_config.py`. I'd keep the rest.

I think we can add a 'slim' that e.g. doesn't include node and CHP, which is what I think ~all container deployments should use, but I definitely don't want to propose adding any images before getting this all set up.

**Comment by manics** [comment]
> Nice! What would be the process for doing e.g. a security patch for the 5.x image once 6.x is being published? Just merge into a PR to a 5.x branch?

Yes. I've pushed a 4.x branch as an example:
https://github.com/manics/jupyterhub-container-images/compare/main..4.x
https://quay.io/repository/manics/jupyterhub-image-test?tab=tags

![Image](https://github.com/user-attachments/assets/1025c66e-3fe3-4e4e-8108-29b2006a2a28)

**Comment by minrk** [comment]
That's really awesome! I'd like to start work toward releasing 5.3 (with #5022 and #5012), and maybe we can try to exercise this as part of it. Does that sound good to you?

**Comment by manics** [comment]
I've transferred the two repos. If everything's working the last step is to create new secrets
- `QUAY_USERNAME`
- `QUAY_PASSWORD`
- `DOCKERHUB_USERNAME`
- `DOCKERHUB_TOKEN`

in https://github.com/jupyterhub/jupyterhub-container-images (and to delete the old ones in this repo). I think we should wait until 5.3.0 is ready so we don't accidentally overwrite the existing 5.2.1 image- there's no `5.2.1-0` at the moment so if we rebuild the image with the new repo the current `5.2.1` will be replaced.

**Comment by minrk** [comment]
Docker builds are stopped in #5024, I'll delete the secrets from this repo when that's merged.

**Comment by minrk** [comment]
The new repo is working great (thanks @manics!) and will publish the 5.3 images. It has already published 5.3.0rc0.
